front50-core/src/main/java/com/netflix/spinnaker/front50/model/plugin/PluginRepository.java

```diff
@@ -0,0 +1,96 @@
+package com.netflix.spinnaker.front50.model.plugin;
+
+import java.util.List;
+
+public interface PluginRepository<T> {
+  /**
+   * Inserts or updates the specified plugin.
+   *
+   * @param data The plugin data to store
+   */
+  void upsert(UpsertData<T> data);
+
+  /**
+   * Returns the plugin by the specified name.
+   *
+   * @param criteria The criteria by which to get the plugin.
+   * @return The plugin object.
+   */
+  T get(GetCriteria criteria);
+
+  /**
+   * Get all plugins based on the criteria.
+   *
+   * @param criteria The criteria by which to filter plugins by.
+   * @return A list of matching plugins.
+   */
+  List<T> list(ListCriteria criteria);
+
+  class GetCriteria {
```

Comment:
nit: I think from an ergonomics perspective, this these criteria objects should use a builder pattern.